### ISSUE TITLE
Network Layer Example

### ISSUE THREAD

**Issue by mertnesvat** [Moya/Moya#399]
I couldn't understand network layer in the examples of moya ([here](https://github.com/Moya/Moya/blob/6a437938c4ac3ae1d6ffce161cfb1fb3c78c67b9/docs/Examples.md) )

I tried to create my own network layer but from example, 
```
struct Network {
    static let provider = MoyaProvider(endpointClosure: endpointClosure)

    static func request(
        target: Github,
        success successCallback: (JSON) -> Void,
        error errorCallback: (statusCode: Int) -> Void,
        failure failureCallback: (Moya.Error) -> Void
    ) {
        provider.request(target) { result in
            switch result {
            case let .Success(response):
                do {
                    try response.filterSuccessfulStatusCodes()
                    let json = try JSON(response.mapJSON())
                    successCallback(json)
                }
                catch error {
                    errorCallback(error)
                }
            case let .Failure(error):
                if target.shouldRetry {
                    retryWhenReachable(target, successCallback, errorCallback, failureCallback)
                }
                else {
                    failureCallback(error)
                }
            }
        }
    }
}

// usage:
Network.request(.Zen, success: { zen in
    showMessage(zen)
}, error: { err in
    showError(err)
}, failure: { _ in
    // oh well, no network apparently
})
```

I use object mapper to map objects and Moya_ObjectMapper with using moya. I use moya like this, 

```
ApiaryProvider.request(Apiary.Home) { (result) -> () in
            
            var success = true
            var message = "Unable to fetch results"

            
            
            switch result {
            case let .Success(response):

                do {
                    var h : HomeRoot = try! response.mapObjectMapper()
                    self.home = h
                    self.loadData()
                    break;

                }catch{
                    
                }
                
            case let .Failure(error):
                guard let error = error as? CustomStringConvertible else {
                    break
                }
                message = error.description
                success = false
            }
        }
```

But I'm using this code block for every different request. So my question is how can I improve this to add extra network layer in every request? or what is the best practice?  

Thanks in advance..

**Comment by ashfurrow** [comment]
Hmm, tricky question – you've got a lot going on here, and I want to make sure I understand your question first. 

Looks like you have a `Network` struct on top of Moya so that you can handle retrying? That's pretty cool, I've done a similar approach [here](https://github.com/artsy/eidolon/blob/24e36a69bbafb4ef6dbe4d98b575ceb4e1d8345f/Kiosk/App/Networking/Networking.swift#L22-L31). Then it looks like you're using ObjectMapper to parse the response data, right? Have you looked at [this library](https://github.com/ivanbruel/Moya-ObjectMapper), which has bindings for Moya and ObjectMapper? It should make it so you don't have to do this every time.

**Comment by mertnesvat** [comment]
Yes you're right I'm using Moya-ObjectMapper but what I couldn't do is creating a network layer to gather all responses in one place. So as you say I need a Network struct on top of Moya, but I have problems with generic parameters T. 
My unfinished network layer is [here](https://gist.github.com/mertnesvat/c98d65c4d9c357ca1ade) problem is when I try to map object returns empty.

**Comment by sunshinejr** [comment]
The problem with your Layer is in this line:
```swift 
var a : T = try! response.mapObjectMapper()
```

Method used in the `Moya-ObjectMapper` bridge has kinda different declaration, without the `Mapper` suffix. I tested your code with change in the mapping part, specifically:
```swift
var a: T = try! response.mapObject()
```

aaaand it works. So if this is your custom method, make sure you test it correctly or just use the method from bridge library. Hope that now you will resolve it!

**Comment by mertnesvat** [comment]
@sunshinejr thanks for your answer but I couldn't find mapObject() method? Which library has it?

**Comment by sunshinejr** [comment]
[This one](https://github.com/ivanbruel/Moya-ObjectMapper) that @ashfurrow was talking about. In README it has only `mapArray()` in examples, but [here](https://github.com/ivanbruel/Moya-ObjectMapper/blob/master/Source/Response%2BObjectMapper.swift) you can see that there are 2 methods, `mapObject()` and `mapArray()`.

**Comment by mertnesvat** [comment]
@sunshinejr @ashfurrow  your're life saver many thanks :) I wasn't aware of the new updates at Moya-ObjectMapper, I implemented it when it was **1.0.0** version and now it is **1.1.2 **so I typed pod update and its working now. :+1:
